src/terminal/input/terminalInput.cpp

```diff
@@ -515,169 +162,442 @@ TerminalInput::OutputType TerminalInput::HandleKey(const INPUT_RECORD& event)
     }
 
-    // Many keyboard layouts have an AltGr key, which makes widely used characters accessible.
-    // For instance on a German keyboard layout "[" is written by pressing AltGr+8.
-    // Furthermore Ctrl+Alt is traditionally treated as an alternative way to AltGr by Windows.
-    // When AltGr is pressed, the caller needs to make sure to send us a pretranslated character in uChar.UnicodeChar.
-    // --> Strip out the AltGr flags, in order for us to not step into the Alt/Ctrl conditions below.
-    if (WI_AreAllFlagsSet(keyEvent.dwControlKeyState, LEFT_CTRL_PRESSED | RIGHT_ALT_PRESSED))
+    const auto ctrlIsPressed = WI_IsAnyFlagSet(controlKeyState, CTRL_PRESSED);
+    const auto ctrlIsReallyPressed = leftCtrlIsReallyPressed || WI_IsFlagSet(controlKeyState, RIGHT_CTRL_PRESSED);
+    const auto shiftIsPressed = WI_IsFlagSet(controlKeyState, SHIFT_PRESSED);
+    const auto altIsPressed = WI_IsAnyFlagSet(controlKeyState, ALT_PRESSED);
+    const auto altGrIsPressed = altIsPressed && ctrlIsPressed;
+
+    // If it's a numeric keypad key, and Alt is pressed (but not Ctrl), then
+    // this is an Alt-Numpad composition and we should ignore these keys. The
+    // generated character will be transmitted when the Alt is released.
+    if (virtualKeyCode >= VK_NUMPAD0 && virtualKeyCode <= VK_NUMPAD9 && altIsPressed && !ctrlIsPressed)
     {
-        WI_ClearAllFlags(keyEvent.dwControlKeyState, LEFT_CTRL_PRESSED | RIGHT_ALT_PRESSED);
+        return MakeOutput({});
     }
 
-    // The Alt modifier initiates a so called "escape sequence".
-    // See: https://en.wikipedia.org/wiki/ANSI_escape_code#Escape_sequences
-    // See: ECMA-48, section 5.3, http://www.ecma-international.org/publications/standards/Ecma-048.htm
-    //
-    // This section in particular handles Alt+Ctrl combinations though.
-    // The Ctrl modifier causes all of the char code's bits except
-    // for the 5 least significant ones to be zeroed out.
-    if (WI_IsAnyFlagSet(keyEvent.dwControlKeyState, ALT_PRESSED) && WI_IsAnyFlagSet(keyEvent.dwControlKeyState, CTRL_PRESSED))
+    // The only enhanced key we care about is the Return key, because that
+    // indicates that it's the key on the numeric keypad, which will transmit
+    // different escape sequences when the Keypad mode is enabled.
+    const auto enhancedReturnKey = WI_IsFlagSet(controlKeyState, ENHANCED_KEY) && virtualKeyCode == VK_RETURN;
+
+    // Using the control key state that we calculated above, combined with the
+    // virtual key code, we've got a unique identifier for the key combination
+    // that we can lookup in our map of predefined key sequences.
+    auto keyCombo = virtualKeyCode;
+    keyCombo += (ctrlIsReallyPressed ? Ctrl : Unmodified);
+    keyCombo += (altIsPressed ? Alt : Unmodified);
+    keyCombo += (shiftIsPressed ? Shift : Unmodified);
+    keyCombo += (enhancedReturnKey ? Enhanced : Unmodified);
+    auto keySequence = _keyMap[keyCombo];
+    if (!keySequence.empty())

```

Comment:
Would it make sense to use `.find()` here instead?

Reply:
Yeah, that should definitely have been a find. Thanks for catching that.